dpkg-testdeb.8
.TH dpkg-testdeb 8 "Release 1.0" "LLNL" "dpkg-testdeb"
.SH NAME
dpkg-testdeb \- Install and test deb packages standalone
.SH SYNOPSIS
.B dpkg-testdeb
.I [OPTIONS]
.I "package [package]..."
.SH DESCRIPTION
.B dpkg-testdeb
is part of dpkg-scripts, a collection of utilities for building
/usr/local-based deb packages.
.LP
.B dpkg-testdeb
allows verification of a single deb package by installing it and
its dependencies inside a temporary /usr/local. 
\fBdpkg-runtests\fR(8) is then invoked on the installed package for 
a simple method to verify that the package works, and has properly 
listed all run-time dependencies. \fBdpkg-verify\fR(8) may also be
optionally run on the installed pacakge, though its use is debatable 
on a freshly installed package.

.SH "OPTIONS"
.TP
.BI "-h, --help"
Display a summary of command-line options.
.TP
.BI "-v, --verbose"
Run \fBdpkg-testdeb\fR in verbose mode.
.TP
.BI "-V, --verify"
Run \fBdpkg-verify\fR(8) in addition to \fBdpkg-runtests\fR(8).

.SH EXAMPLES
.nf
 > dpkg-testdeb ifort-10.1.011_3_linux-amd64.deb

 dpkg-testdeb: Installing ifort-10.1.011 in temporary /usr/local...
 dpkg-testdeb: Fixing up dependencies...
 dpkg-testdeb: Running dpkg-runtests on installed packages...
 dpkg-runtests: ifort-10.1.011                           ok
 dpkg-runtests: Summary: all tests passed
 dpkg-testdeb: Success
.fi

.SH "SEE ALSO"
.BR dpkg-runtests (8), 
.BR	dpkg-verify (8),
.BR dpkg-tmplocal (8)


NEWS
========================================================================
Release Notes for dpkg-scripts-1.66                          17 Mar 2009
========================================================================

* Fix dpkg-tmplocal option processing to stop at the first non-option
  argument.

========================================================================
Release Notes for dpkg-scripts-1.65                          16 Mar 2009
========================================================================

* New dpkg-tmplocal options:
   -v (--verbose)   Report each step as it is performed.
   -p (--preserve)  Do not clean up temporary directory.
   -d (--dir)       Bind a pre-existing directory on /usr/local (implies -p)
   -o (--slash-opt) Bind on /opt instead of /usr/local.

* Avoid overriding PATH in dpkg-wrap [chaos bug 454]

========================================================================
Release Notes for dpkg-scripts-1.64                          22 Apr 2008
========================================================================

* Do not hide default package dotkits from "use -l" by default.

* Add 'dkhidedefault' package flag to hide default package dotkit.

* Add 'dkhidemain' package flag to hide main package dotkit.
  The 'dkhide' package flag hides both main and default packaged dotkits.

========================================================================
Release Notes for dpkg-scripts-1.63                          20 Apr 2008
========================================================================

* Fix bug in linkdata2 generation by dpkg-mkdeb where long symlink 
  targets were encoded into base64 with line breaks.

* Rewrite dpkg-verify in C for efficiency.  Work around linkdata2 files 
  damaged as described above.

* The dpkg-nonpackaged utility is replaced with dpkg-verify --unpackaged.

========================================================================
Release Notes for dpkg-scripts-1.62                          10 Apr 2008
========================================================================

* Added dpkg-mkdeb options: '-q basename', '-q basever', and -m metafile.
  Fixed a bug in dpkg-mkdeb -q default.

========================================================================
Release Notes for dpkg-scripts-1.61 (includes .59, .60)      09 Apr 2008
========================================================================

* Change '-q names' query to return only non-default packages names, 
  and add '-q default' package to return default package names.

* Add regression test.

========================================================================
Release Notes for dpkg-scripts-1.58                          20 Mar 2008
========================================================================

* Add 'verboselog' flag to allow logging of arguments passed to wrapped
  commands to be turned on for some packages.

========================================================================
Release Notes for dpkg-scripts-1.57                          28 Feb 2008
========================================================================

* Drop logging of arguments passed to wrapped commands.

========================================================================
Release Notes for dpkg-scripts-1.56                          21 Feb 2008
========================================================================

* Change hash-bang in all scripts to #!/bin/bash --posix to avoid
  high shell startup overhead due to LC BASH_ENV setting.
  Also make a couple of executable paths fully qualified in the 
  dpkg-wrap call chain to avoid gratuitous PATH walks.

* New dpkg-defaults script for listing package defaults.

========================================================================
Release Notes for dpkg-scripts-1.55                          15 Feb 2008
========================================================================

* Add dpkg-mkdeb -q names query to list all possible package names,
  needed on build farm.

* On failure, dpkg-mkdeb should not dump logfile to stderr if -v is 
  already sending it there.

* Do not build with 'notmproot' behavior by default.  Instead, populate
  a private /usr/local for PKG_BUILDREQUIRES, but install to a regular
  tmproot.  The -t option allows the system /usr/local to be used.

* Change the dpkg-docreport color scheme to white and gray.

========================================================================
Release Notes for dpkg-scripts-1.54                          12 Feb 2008
========================================================================

* Support for symlinking lcc plugins in default pacakges.

* New macros for use in package.conf: 
  $lccdir - location of lcc plugin directory
  $pkgname - package name (after subpackage processing)

* Ensure file permissions are such that deb can be installed by non-root
  user.

* Define $pkg_default (set to PKG_DEFAULT value) in package.macros
  metadata file.

* Handle some corner cases in PKG_DEFAULT parsing.

========================================================================
Release Notes for dpkg-scripts-1.53                          01 Feb 2008
========================================================================

* Add 'dkhide' flag to generate #a in dotkit, which hides it from
  the use -l listing.  Generate #a in default packages with or without
  'dkhide' flag.

* Add ${dotkitdir} macro which can be used instead of hard coding
  /usr/local/etc/dotkit in package.conf's.

* [dpkg-mkdeb] Make unknown entries in PKG_FLAGS fatal.
  
========================================================================
Release Notes for dpkg-scripts-1.52                          30 Jan 2008
========================================================================

* Add 'nocheckwrap' flag to disable checking for existence of wrapped 
  executables (e.g. to support wrapping of unpackaged files).


dpkg-testdeb.sh
#!/bin/bash --posix

############################################################################
# Copyright (C) 2007 Lawrence Livermore National Security, LLC
# Produced at Lawrence Livermore National Laboratory.
# Written by Jim Garlick <garlick@llnl.gov>.
# UCRL-CODE-235516
# 
# This file is part of dpkg-scripts, a set of utilities for managing 
# packages in /usr/local with dpkg.
# 
# dpkg-scripts is free software; you can redistribute it and/or modify it 
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2 of the License, or (at your option)
# any later version. 
#
# dpkg-scripts is distributed in the hope that it will be useful, but WITHOUT 
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for 
# more details.
#
# You should have received a copy of the GNU General Public License along
# with dpkg-scripts; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
############################################################################

# constants
declare -r prog=`basename $0`
declare -r dpkgroot=/usr/local
declare -r dpkginit=/usr/sbin/dpkg-initialize
declare -r usage="\
Usage: ${prog} [OPTIONS]... PKG [PKG]...\n\
    -h, --help      Display this message.\n\
    -v, --verbose   Verbose output.\n\
    -V, --verify    Also run dpkg-verify against package(s)."

# user-supplied options
tmplocal=0
verify=0
verbose=0

############################################################################
#  Functions
############################################################################

die()
{
	echo  -e "${prog}: $@" >&2
    exit 1
}

log_msg()
{
    echo -e "${prog}: $@" >&2
}

log_verbose()
{
    [ $verbose = 1 ] && log_msg "$@"
}

package_name_list()
{
    local pkgs
    for file in $@; do
        pkgs="${pkgs:+$pkgs }$(dpkg-deb --field ${file} Package)" 
        [ $? -ne 0 ] && die "dpkg-deb on ${file} failed."
    done
    echo "$pkgs"
}

runcmd()
{
    local vopt=$verbose
    while getopts v opt; do
        case ${opt} in
            v) vopt=1 ;;
            *) die "runcmd internal error: Bad option: -${opt}" ;;
        esac
    done
    shift $((${OPTIND} - 1))

    log_verbose "Running $@"
    if [ $vopt = 1 ]; then
       $@ 
    else
       $@ >/dev/null 2>&1
    fi
}

############################################################################
#  Main script
############################################################################

# abort early if root
[ `id -u` = 0 ] && die "Do not run as root."

# process options
saved_args=$@
args=`/usr/bin/getopt -uo htVv -l help,tmplocal,verbose,verify -n ${prog} -- $@`
if [ $? -ne 0 ]; then
   die "$usage"
fi

set -- ${args}
while true; do
    case "$1" in
     -h|--help)     echo -e "$usage"; exit 0;;
	 -t|--tmplocal) tmplocal=1; shift;; 
     -V|--verify)   verify=1; shift;;
     -v|--verbose)  verbose=1; shift;;
	 --)            shift; break;;
	 *)             die "Internal error: option '$1' not understood" ;;
	esac
done

[ $#  -gt 0 ] ||  die "Must supply a package to test.\n$usage"

# 
#  If not running in tmplocal reinvoke under dpkg-tmplocal(8)
#
if [ $tmplocal = 0 ]; then
   exec dpkg-tmplocal -- $0 --tmplocal ${saved_args}
fi

#
#  Initialize tmplocal
#
$dpkginit || die "Failed to initialize tmp /usr/local. Did dpkg-tmplocal fail?"

#
#  Get package names for later use in dpkg-runtests
#
packages=$(package_name_list "$@") || die "Failed to read all packages"
log_msg "Installing ${packages} in temporary /usr/local..."

# 
#  Run dpkg to install packages. Expect failure here as there may
#   be missing dependencies which apt-get -f install will fix up.
# 
PATH=/sbin:/usr/sbin:$PATH runcmd dpkg --force-not-root -i "$@" 

#
#  Run 'apt-get -f install'
#
log_msg "Fixing up dependencies..."
runcmd apt-userinst -f               || die "apt-userinst failed"

#
#  Run tests for installed packages
#
log_msg "Running dpkg-runtests on installed packages..."
runcmd -v dpkg-runtests ${verbose+-v }${packages}  || die "dpkg-runtests failed"

#
#  Run dpkg-verify if requested
#
if [ $verify = 1 ]; then
   log_msg "Running dpkg-verify on installed packages..."
   runcmd -v dpkg-verify ${packages} || die "dpk-verify failed"
fi

log_msg "Success"
exit 0

# vi: ts=4 sw=4 expandtab
